zebra/zebra_mpls.c

```diff
@@ -135,155 +165,615 @@ mpls_processq_init (struct zebra_t *zebra);
+ * NHLFEs are removed, the entire entry is deleted.
  */
 static int
-label_cmp (const void *p1, const void *p2)
+lsp_uninstall (struct zebra_vrf *zvrf, mpls_label_t label)
 {
-  const zebra_ile_t *ile1 = p1;
-  const zebra_ile_t *ile2 = p2;
+  struct hash *lsp_table;
+  zebra_ile_t tmp_ile;
+  zebra_lsp_t *lsp;
+  zebra_nhlfe_t *nhlfe, *nhlfe_next;
+  char buf[BUFSIZ];
 
-  return (ile1->in_label == ile2->in_label);
+  /* Lookup table. */
+  lsp_table = zvrf->lsp_table;
+  if (!lsp_table)
+    return -1;
+
+  /* If entry is not present, exit. */
+  tmp_ile.in_label = label;
+  lsp = hash_lookup (lsp_table, &tmp_ile);
+  if (!lsp || !lsp->nhlfe_list)
+    return 0;
+
+  /* Mark NHLFEs for delete or directly delete, as appropriate. */
+  for (nhlfe = lsp->nhlfe_list; nhlfe; nhlfe = nhlfe_next)
+    {
+      nhlfe_next = nhlfe->next;
+
+      /* Skip static NHLFEs */
+      if (nhlfe->type == ZEBRA_LSP_STATIC)
+        continue;
+
+      if (IS_ZEBRA_DEBUG_MPLS)
+        {
+          nhlfe2str (nhlfe, buf, BUFSIZ);
+          zlog_debug ("Del LSP in-label %u type %d nexthop %s flags 0x%x",
+                      label, nhlfe->type, buf, nhlfe->flags);
+        }
+
+      if (CHECK_FLAG(nhlfe->flags, NHLFE_FLAG_SELECTED))
+        {
+          UNSET_FLAG (nhlfe->flags, NHLFE_FLAG_CHANGED);
+          SET_FLAG (nhlfe->flags, NHLFE_FLAG_DELETED);
+        }
+      else
+        {
+          nhlfe_del (nhlfe);
+        }
+    }
+
+  /* Queue LSP for processing, if needed, else delete. */
+  if (CHECK_FLAG(lsp->flags, LSP_FLAG_INSTALLED))
+    {
+      if (lsp_processq_add (lsp))
+        return -1;
+    }
+  else if (!lsp->nhlfe_list &&
+           !CHECK_FLAG (lsp->flags, LSP_FLAG_SCHEDULED))
+    {
+      if (IS_ZEBRA_DEBUG_MPLS)
+        zlog_debug ("Del LSP in-label %u flags 0x%x",
+                    lsp->ile.in_label, lsp->flags);
+
+      lsp = hash_release(lsp_table, &lsp->ile);
+      if (lsp)
+        XFREE(MTYPE_LSP, lsp);
+    }
+
+  return 0;
 }
 
 /*
- * Check if an IPv4 nexthop for a NHLFE is active. Update nexthop based on
- * the passed flag.
- * NOTE: Looking only for connected routes right now.
+ * This function is invoked upon change to label block configuration; it
+ * will walk all registered FECs with label-index and appropriately update
+ * their local labels and trigger client updates.
  */
-static int
-nhlfe_nexthop_active_ipv4 (zebra_nhlfe_t *nhlfe, struct nexthop *nexthop)
+static void
+fec_evaluate (struct zebra_vrf *zvrf, int add)
 {
-  struct route_table *table;
-  struct prefix_ipv4 p;
   struct route_node *rn;
-  struct rib *match;
-  struct nexthop *match_nh;
+  zebra_fec_t *fec;
+  u_int32_t old_label, new_label;
+  int af;
+  char buf[BUFSIZ];
 
-  table = zebra_vrf_table (AFI_IP, SAFI_UNICAST, VRF_DEFAULT);
-  if (!table)
-    return 0;
+  for (af = AFI_IP; af < AFI_MAX; af++)
+    {
+      for (rn = route_top(zvrf->fec_table[af]); rn; rn = route_next(rn))
```

Comment:
zebra is segfaulting here:
```
 debian(config)# mpls label global-block 16 500
 vtysh: error reading from zebra: Success (0)Warning: closing connection to zebra because of an I/O error!
```
You need to fix the loop like this:
`for (af = AFI_IP; af <= AFI_IP6; af++)`